### PR TITLE
fix(gatsby): handle case of removing trailing slash in inc builds

### PR DESCRIPTION
**Pull request by pieh**
## Description

This is to handle cases like:

```js
exports.onCreatePage = ({ page, actions }) => {
  if (someConditionThatCheckPagePath) {
    actions.deletePage(page)
    actions.createPage({
      ...page,
      path: slightlyChangedPagePath,
    })
  }
}
```

Used in plugins like `gatsby-plugin-remove-trailing-slashes` etc.

For now this add failing test scenario, will add a fix for it in next commit

## Related Issues

Fixes https://github.com/gatsbyjs/gatsby/issues/29946

### CODE REVIEW

File: packages/gatsby/src/commands/build-utils.ts [hunk 1]
```diff
@@ -66,11 +66,84 @@ export const removePageFiles = async (
...
+// almost c&p from `cache-dir/normalize-page-path
+function normalizePagePath(path: string): string {
+  if (path === `/`) {
+    return `/`
+  }
+  if (path.charAt(path.length - 1) === `/`) {
+    return path.slice(0, -1)
+  }
+  return path
+}
```
Comment (pieh) [inline packages/gatsby/src/commands/build-utils.ts:78]: do we have something like this already in Node.js portion of code? I did few greps and couldn't find one :(

**Review by wardpeet — DISMISSED** [review]
Code looks good, would it make sense to add unit tests to packages/gatsby/src/commands/build-utils.ts to test the logic instead of only relying on e2e tests

File: packages/gatsby/src/commands/__tests__/built-utils.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    })
+
+    const results = calcDirtyHtmlFiles(state)
+    expect(results).toMatchInlineSnapshot(`
+      Object {
+        "toCleanupFromTrackedState": Set {},
+        "toDelete": Array [],
+        "toRegenerate": Array [],
+      }
+    `)
```
Comment (pieh) [inline packages/gatsby/src/commands/__tests__/built-utils.ts:79]: I will adjust tests to not use snapshots as that's often foot gun - I just pushed current state of those

**Review by vladar — APPROVED** [review]
Thank you 👍

**Comment by vladar** [comment]
Published in `gatsby@3.0.2`

**Pull request merged**